astropy/modeling/fitting.py

```diff
@@ -610,35 +624,132 @@ def __call__(self, model, x, y, z=None, weights=None, **kwargs):
             Fitted model after outlier removal.
         """
 
-        fitted_model = self.fitter(model, x, y, z, weights=weights, **kwargs)
-        filtered_weights = weights
+        # For single models, the data get filtered here at each iteration and
+        # then passed to the fitter, which is the historical behaviour and
+        # works even for fitters that don't understand masked arrays. For model
+        # sets, the fitter must be able to filter masked data internally,
+        # because fitters require a single set of x/y co-ordinates whereas the
+        # eliminated points can vary between models. To avoid this limitation,
+        # we could fall back to looping over individual model fits, but it
+        # would likely be fiddly and involve even more overhead (and the
+        # non-linear fitters don't work with model sets anyway, as of writing).
+
+        if len(model) == 1:
+            model_set_axis = None
+        else:
+            if not hasattr(self.fitter, 'supports_masked_input') or \
+               self.fitter.supports_masked_input is not True:
+                raise ValueError("{0} cannot fit model sets with masked "
+                                 "values".format(type(self.fitter).__name__))
+
+            # The intention appears to be that fitters use their input model's
+            # model_set_axis to determine how their input data are stacked:
```

Comment:
After the latest changes related to this in `LinearLSQFitter` , it's not an "intention" any more and this comment may be confusing for readers later. Is it better to just state this, e.g.  `Fitters use their ``model_set_axis`` to stack input data`.

Reply:
But fitters don't have a `model_set_axis`, do they? Perhaps they should have, for consistency with the models, but that's probably a whole PR of its own. Or is that not what you meant? We can certainly clarify the comment though, if the intention is now better defined.

Reply:
All I meant is just clarify the comment. We know for a fact that `model.model_set_axis` is used to stack the input. Let's make it clear in the comment. Obviously this is a very minor point.

I don't see a compelling need for `fitter.set_axis` at the moment.

Reply:
So just remove "The intention appears to be that"?

Reply:
:+1: